src/BSON/Document.c

```diff
@@ -192,6 +192,36 @@ static bool php_phongo_document_get(php_phongo_document_t* intern, char* key, si
 	return true;
 }
 
+static bool php_phongo_document_get_by_zval(php_phongo_document_t* intern, zval* key, zval* return_value, bool null_if_missing)
```

Comment:
I created two `by_zval` variants for `get` and `has`, which are used by the ArrayAccess methods and dimension handlers. These both abstraction the `zval_try_get_tmp_string` logic, which was necessary to avoid possibly modifying the zval argument with a basic cast (see: [Casts and Operations](https://www.phpinternalsbook.com/php7/zvals/casts_and_operations.html)).

I also added a test for this using an assigned variable instead of a literal to assert that we don't modify anything inadvertently.